### PR TITLE
Store `plain_rewritable` disk metadata in the separate `'__meta'` layout; add flat object storage directory structure

### PR DESCRIPTION
**Pull request by jkartseva**
An example of object_storage layout after the change:
```
./__meta/ekqksqmqjootkiisofrgxanqigdzlfbz/prefix.path/xl.meta
./__meta/llhxeztcwqqfrwzjzqebtxnqpcortgor/prefix.path/xl.meta
./__meta/zaflawfkpuswindcdoocfqdglaplgcut/prefix.path/xl.meta
./__meta/fblzzufsqilwkhgovbkhzjaceevpcwql/prefix.path/xl.meta
./__meta/wxbhorxekhnmgqbvhhxexecjletwwpes/prefix.path/xl.meta
./__meta/cjznxqrydozehsabrzboqqtxpcaiyepw/prefix.path/xl.meta
./__meta/pmazlwpxffbtyozzpiabbloylpmiqlet/prefix.path/xl.meta
./__meta/sxpcgyzjzgwzwwsmjvufgsgigjkfmbec/prefix.path/xl.meta
./__meta/rvwbziebhbxqbixduvewyqrjlanylkel/prefix.path/xl.meta
./__meta/mmrsmwpdcxirjagpygfpdcpebnepzgdm/prefix.path/xl.meta
./__meta/rphturfvrvvycmrggucaqypenfqlzazn/prefix.path/xl.meta
./__meta/edtkygitaarrxizdsigsaatvhtvuctyc/prefix.path/xl.meta
./__meta/gtfiqveyrtdhxkmwltiswtyvnhovpntq/prefix.path/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/serialization.json/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/data.bin/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/primary.cidx/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/columns.txt/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/checksums.txt/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/data.cmrk3/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/metadata_version.txt/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/default_compression_codec.txt/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/partition.dat/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/count.txt/xl.meta
./ekqksqmqjootkiisofrgxanqigdzlfbz/minmax_a.idx/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/serialization.json/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/data.bin/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/primary.cidx/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/columns.txt/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/checksums.txt/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/data.cmrk3/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/metadata_version.txt/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/default_compression_codec.txt/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/partition.dat/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/count.txt/xl.meta
./llhxeztcwqqfrwzjzqebtxnqpcortgor/minmax_a.idx/xl.meta

```
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Store the `plain_rewritable` disk directory metadata in `__meta` layout, separately from the merge tree data in the object storage. Move the `plain_rewritable` disk to a flat directory structure.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit f06f2292035efb819c9d53e9315cc6331b11417e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65751/f06f2292035efb819c9d53e9315cc6331b11417e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by alexey-milovidov** [comment]
Should we also make the directory structure flat - avoid extracting common prefixes?
Because the prefixes that appear common can be not common later.

**Comment by jkartseva** [comment]
@alexey-milovidov
MergeTree requires traversing direct subdirectories, hence, we'll need to get the information that B is a child of A.

With a flat directory structure, as I understand it, each directory will have a unique pseudo-random name, possibly with a prefix reflecting the depth of the MT path (if https://github.com/ClickHouse/ClickHouse/issues/65321 is implemented). Thus, the information that 'B is a child of A' will not be reflected in the flat structure paths.
I mean something like:
```
001_bbvxygjiwmoapsdeshsggmqcgotsdbon
002_twkwwcncwfoufzuhoalhqszglxmqtcji
```

We have an in-memory map with the virtual MT paths (e.g., 'store/cbf/cbfff232-0df0-48f2-86b5-5bb39c897173/4_6_6_1/') as keys and the object storage paths as values. If it's okay to use this map as the source of truth for directories' direct children, I can make the structure flat.
___
Alternatively, S3 paths may be in a form of `([a-z]{16}\/){0,3}[a-z]{16}`, i.e. a pseudo-random parent path is a prefix of child's path, an example:
```
xfyrnptgybuamjjt
xfyrnptgybuamjjt/zxfhacytunbmxtuk
xfyrnptgybuamjjt/zxfhacytunbmxtuk/oalhqszglxmqtcji
```
But, this is not a 'flat' structure.

**Comment by alexey-milovidov** [comment]
> If it's okay to use this map as the source of truth for directories' direct children, I can make the structure flat.

Yes, this is what I've meant: use a flat structure (where all directories are near each other) and reconstruct the directory hierarchy from the metadata files.

**Comment by jkartseva** [comment]
> PS. About adding the depth in the name - probably it does not make sense after that.

Since the regular MergeTree file paths won't be stored in the in-memory map, I think it still makes sense.

**Comment by Zawa-ll** [comment]
Sorry for adding a comments even though its not that relevant to your changes here...

I am also encountering failures in the `00002_log_and_exception_messages_formatting` test. This error is not relevant to my changes, but for merging my changes, it needs to be passed.

It seems that the failure is due to a known flaky test, as discussed in issue #66401. Someone mentioned that a fix is being worked on in pull request #66406. 
But I wasn't sure, should I wait for that one to be fixed and then rebase my branch and retry the tests once the fix is merged?

**Comment by alexey-milovidov** [comment]
Yes, flaky tests are strict blockers for merging, and fixing them have the highest priority.

**Comment by jkartseva** [comment]
@Zawa-ll you can merge the https://github.com/ClickHouse/ClickHouse/pull/66406 branch into your branch if you want the see the test results before that PR is merged.

**Comment by Zawa-ll** [comment]
> @Zawa-ll you can merge the #66406 branch into your branch if you want the see the test results before that PR is merged.

Sure, thank you for your response! This really helps me out!

**Comment by jkartseva** [comment]
Changes: 
Storing the metadata (prefix.path file) behind its own ‘__meta’ key prefix.
ObjectStorageKeyGenerator::generate signature is changed to carry an optional `key_prefix` parameter to override the storage key prefix.
The in-memory map keys do not end with the trailing `'/'`.
For the backward compatibility, if the disk already has data and `__meta` key is not among it, use the key generator based on the common prefix.
Flat directory structure key generator.
The `listDirectory` is changed to rely on the in-memory map for listing the subdirectories (the flat structure goes not preserve hierarchy).
The in-memory map comparator is changed so it compares paths by the number of hops, then lexicographically. This is to find subdirectories quickly.
Flat directory structure is enabled only for newly created disks; if metadata is stored in a separate layout, the disk also utilizes flat structure.

File: src/Disks/ObjectStorages/MetadataStorageFromPlainRewritableObjectStorage.cpp [hunk 1]
```diff
@@ -149,51 +168,100 @@ std::vector<std::string> getDirectChildrenOnRewritableDisk(
...
+        auto key_prefix = object_storage->generateObjectKeyForPath(path, getMetadataKeyPrefix()).serialize();
+        return object_storage->existsOrHasAnyChild(key_prefix);
+    }
+
+    return false;
+}
+
+bool MetadataStorageFromPlainRewritableObjectStorage::isDirectory(const std::string & path) const
+{
+    if (getMetadataKeyPrefix() != object_storage->getCommonKeyPrefix())
```
Comment (kssenii) [inline src/Disks/ObjectStorages/MetadataStorageFromPlainRewritableObjectStorage.cpp:225]: May be move `getMetadataKeyPrefix() != object_storage->getCommonKeyPrefix()` into separate method named `useSeparateLayoutForMetadata`? It is used quite often

File: src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp [hunk 2]
```diff
@@ -20,38 +21,57 @@ namespace
...
         /* settings */ {});
 
     write_created = true;
 
-    [[maybe_unused]] auto result = path_map.emplace(path, std::move(key_prefix));
-    chassert(result.second);
+    {
+        std::unique_lock lock(mutex);
+        auto & map = path_map.map;
+        [[maybe_unused]] auto result = map.emplace(path.parent_path(), std::move(key_prefix));
```
Comment (kssenii) [inline src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp:72]: why take `path.parent_path()` here and not just `path`? `path` should already be a directory we want (as this is CreateDirectoryOperation) or I am misunderstanding?

File: src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp [hunk 3]
```diff
@@ -20,38 +21,57 @@ namespace
...
 
-    auto object_key = ObjectStorageKey::createAsRelative(key_prefix, PREFIX_PATH_FILE_NAME);
+    LOG_TRACE(
+        getLogger("MetadataStorageFromPlainObjectStorageCreateDirectoryOperation"),
+        "Creating metadata for directory '{}' with remote path='{}'",
+        path,
+        metadata_object_key.serialize());
 
-    auto object = StoredObject(object_key.serialize(), path / PREFIX_PATH_FILE_NAME);
+    auto metadata_object = StoredObject(metadata_object_key.serialize(), path / PREFIX_PATH_FILE_NAME);
```
Comment (kssenii) [inline src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp:59]:
```suggestion
    auto metadata_object = StoredObject(/* remote_path */metadata_object_key.serialize(), /* local_path */path / PREFIX_PATH_FILE_NAME);
```

File: src/Disks/ObjectStorages/InMemoryPathMap.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        bool operator()(const std::filesystem::path & path1, const std::filesystem::path & path2) const
+        {
+            auto d1 = std::distance(path1.begin(), path1.end());
+            auto d2 = std::distance(path2.begin(), path2.end());
+            if (d1 != d2)
+                return d1 < d2;
+            return path1 < path2;
+        }
+    };
+    using Map = std::map<std::filesystem::path, std::string, PathComparator>;
```
Comment (kssenii) [inline src/Disks/ObjectStorages/InMemoryPathMap.h:24]:
```suggestion
    /// Local -> Remote path.
    using Map = std::map<std::filesystem::path, std::string, PathComparator>;
```

File: src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp (hunk 2 above)
Comment (jkartseva) [reply src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp:72]:
```
std::cout << std::filesystem::path("/path/to/file/").parent_path();
"/path/to/file"
```
It's just a weird way to trim `'/'`. I'll make it more explicit, thanks!

**Review by kssenii — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Something strange with unit tests.

**Comment by jkartseva** [comment]
https://github.com/ClickHouse/ClickHouse/actions/runs/10025260950/job/27711549502?pr=65751

```ALTER TABLE compression_codec_on_alias (ADD COLUMN `c3` ALIAS c2 CODEC(tuple('second', 6, toLowCardinality(6), 6, materialize(6), 6, toUInt256(6), 6, toNullable(6), 6, 6), ZSTD) AFTER c2)```

```
2024-07-21 10:23:24	 #5  0x000000000c3f6f13 in DB::abortOnFailedAssertion (description=..., trace=<optimized out>, trace_offset=trace_offset@entry=0, trace_size=<optimized out>) at /build/src/Common/Exception.cpp:47
2024-07-21 10:23:24	         logger = <optimized out>
2024-07-21 10:23:24	 #6  0x000000000c3f7869 in DB::handle_error_code (msg=..., code=49, trace=..., remote=<optimized out>) at /build/src/Common/Exception.cpp:70
2024-07-21 10:23:24	 No locals.
2024-07-21 10:23:24	 #7  DB::Exception::Exception (this=0x7f02ef320c80, msg_masked=..., code=49, remote_=<optimized out>) at /build/src/Common/Exception.cpp:114
2024-07-21 10:23:24	 No locals.
2024-07-21 10:23:24	 #8  0x0000000006dfc14c in DB::Exception::Exception (this=0x7f02ef320c80, msg=..., code=49, remote_=false) at /build/src/Common/Exception.h:93
2024-07-21 10:23:25	 No locals.
2024-07-21 10:23:25	 #9  DB::Exception::Exception (this=0x7f02ef320c80, msg=..., code=49) at /build/src/Common/Exception.h:64
2024-07-21 10:23:25	 No locals.
2024-07-21 10:23:25	 #10 0x0000000006e2530b in DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&> (this=0x7f02ef320c80, code=49, fmt=..., args=...) at /build/src/Common/Exception.h:111
2024-07-21 10:23:25	 No locals.
2024-07-21 10:23:25	 #11 0x0000000011606fba in DB::executeQueryImpl (begin=begin@entry=0x7f02ef2b83c0 "ALTER TABLE compression_codec_on_alias (ADD COLUMN `c3` ALIAS c2 CODEC(tuple('second', 6, toLowCardinality(6), 6, materialize(6), 6, toUInt256(6), 6, toNullable(6), 6, 6), ZSTD) AFTER c2)", end=0x7f02ef2b847b "", context=..., flags=..., stage=DB::QueryProcessingStage::Complete, istr=0x0) at /build/src/Interpreters/executeQuery.cpp:800

**Comment by jkartseva** [comment]
https://github.com/ClickHouse/ClickHouse/actions/runs/10025260950/job/27711554889?pr=65751#step:8:2640

```
2024-07-21 15:45:27 02497_trace_events_stress_long:                                         Cannot insert coverage data:  [Errno 111] Connection refused
2024-07-21 15:45:27 Cannot insert coverage data:  [Errno 111] Connection refused
2024-07-21 15:45:27 Cannot insert coverage data:  [Errno 111] Connection refused
2024-07-21 15:45:27 Cannot insert coverage data:  [Errno 111] Connection refused
```

**Comment by alexey-milovidov** [comment]
@jkartseva, The problem was introduced in #43639, it appears unsafe. I'm checking it here: https://github.com/ClickHouse/ClickHouse/pull/66804

**Comment by alexey-milovidov** [comment]
Also https://github.com/ClickHouse/ClickHouse/pull/66821

**Pull request merged**